lib/ae_mdw/mnesia.ex

```diff
@@ -0,0 +1,91 @@
+defmodule AeMdw.Mnesia do
+  @moduledoc """
+  Mnesia wrapper to provide a simpler API.
+  """
+
+  @type table() :: atom()
+  @type record() :: tuple()
+  @type key() :: term()
+  @type direction() :: :forward | :backward
+  @type cursor() :: key() | nil
+
+  @end_token :"$end_of_table"
+
+  @spec last_key(table()) :: {:ok, key()} | :none
+  def last_key(tab) do
+    case :mnesia.dirty_last(tab) do
+      @end_token -> {[], nil}
+      last_key -> {:ok, last_key}
+    end
+  end
+
+  @spec fetch(table(), key()) :: {:ok, record()} | :not_found
+  def fetch(tab, key) do
+    case :mnesia.dirty_read(tab, key) do
+      [record] -> {:ok, record}
+      [] -> :not_found
+    end
+  end
+
+  @spec fetch!(table(), key()) :: record()
+  def fetch!(tab, key) do
+    {:ok, record} = fetch(tab, key)
+
+    record
+  end
+
+  @spec fetch_keys(table(), direction(), cursor(), non_neg_integer()) :: {[record()], cursor()}
+  def fetch_keys(tab, :forward, nil, limit) do
+    case :mnesia.dirty_first(tab) do
+      @end_token -> {[], nil}
+      first_key -> fetch_keys(tab, :forward, first_key, limit)
+    end
+  end
+
+  def fetch_keys(tab, :forward, first_key, limit) do
+    {keys, cursor} = fetch_forward_keys(tab, first_key, limit, [])
+
+    {[first_key | Enum.reverse(keys)], cursor}
+  end
+
+  def fetch_keys(tab, :backward, nil, limit) do
+    case :mnesia.dirty_last(tab) do
+      @end_token -> {[], nil}
+      last_key -> fetch_keys(tab, :backward, last_key, limit)
+    end
+  end
+
+  def fetch_keys(tab, :backward, last_key, limit) do
+    {keys, cursor} = fetch_backward_keys(tab, last_key, limit, [])
+
+    {[last_key | Enum.reverse(keys)], cursor}
+  end
+
+  defp fetch_forward_keys(tab, current_key, 1, acc) do
+    case :mnesia.dirty_next(tab, current_key) do
+      @end_token -> {acc, nil}
+      next_key -> {acc, next_key}
+    end
+  end
+
+  defp fetch_forward_keys(tab, current_key, limit, acc) do
+    case :mnesia.dirty_next(tab, current_key) do
+      @end_token -> {acc, nil}
+      next_key -> fetch_forward_keys(tab, next_key, limit - 1, [next_key | acc])
```

Comment:
I 100% agree we could use `:mnesia.select` instead. But for now I let it the same way it was done before because I didn't want to change much or break anything. But this should be done as a subsequent implementation. The good thing is that this will only need to be changed in a single place

Reply:
Done, included everything inside a single transaction